src/algebra/squarefree.lean

```diff
@@ -182,9 +182,134 @@ begin
   simp,
 end
 
-instance : decidable_pred (squarefree : ℕ → Prop)
-| 0 := is_false not_squarefree_zero
-| (n + 1) := decidable_of_iff _ (squarefree_iff_nodup_factors (nat.succ_ne_zero n)).symm
+theorem squarefree_iff_prime_squarefree {n : ℕ} : squarefree n ↔ ∀ x, prime x → ¬ x * x ∣ n :=
+squarefree_iff_irreducible_sq_not_dvd_of_exists_irreducible ⟨_, prime_two⟩
+
+/-- Assuming that `n` has no factors less than `k`, returns the smallest prime `p` such that
+  `p^2 ∣ n`. -/
+def min_sq_fac_aux : ℕ → ℕ → option ℕ
+| n k :=
+  if h : n < k * k then none else
+  have nat.sqrt n + 2 - (k + 2) < nat.sqrt n + 2 - k,
+    by { rw nat.add_sub_add_right, exact nat.min_fac_lemma n k h },
+  if k ∣ n then
+    let n' := n / k in
+    have nat.sqrt n' + 2 - (k + 2) < nat.sqrt n + 2 - k, from
+      lt_of_le_of_lt (nat.sub_le_sub_right
+        (nat.add_le_add_right (nat.sqrt_le_sqrt $ nat.div_le_self _ _) _) _) this,
+    if k ∣ n' then some k else min_sq_fac_aux n' (k + 2)
+  else min_sq_fac_aux n (k + 2)
+using_well_founded {rel_tac :=
+  λ _ _, `[exact ⟨_, measure_wf (λ ⟨n, k⟩, nat.sqrt n + 2 - k)⟩]}
+
+/-- Returns the smallest prime factor `p` of `n` such that `p^2 ∣ n`, or `none` if there is no
+  such `p` (that is, `n` is squarefree). See also `squarefree_iff_min_sq_fac`. -/
+def min_sq_fac (n : ℕ) : option ℕ :=
+if 2 ∣ n then
+  let n' := n / 2 in
+  if 2 ∣ n' then some 2 else min_sq_fac_aux n' 3
+else min_sq_fac_aux n 3
+
+private def min_sq_fac_prop (n : ℕ) : option ℕ → Prop
```

Comment:
This probably shouldn't be private if it appears in the statement of a non-private theorem. At the very least it will look horrible in docs.

```suggestion
def min_sq_fac_prop (n : ℕ) : option ℕ → Prop
```